sources/ghostdog/src/nvme.rs

```diff
@@ -0,0 +1,362 @@
+            vs: [__u8::default(); 1024],
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Default, Debug)]
+/// Represents an NVME command send through ioctl
+struct NvmePassthruCmd {
+    opcode: __u8,
+    flags: __u8,
+    rsvd1: __u16,
+    nsid: __u32,
+    cdw2: __u32,
+    cdw3: __u32,
+    metadata: __u64,
+    addr: __u64,
+    metadata_len: __u32,
+    data_len: __u32,
+    cdw10: __u32,
+    cdw11: __u32,
+    cdw12: __u32,
+    cdw13: __u32,
+    cdw14: __u32,
+    cdw15: __u32,
+    timeout_ms: __u32,
+    result: __u32,
+}
+
+#[derive(Debug)]
+/// Represents an EBS NVME volume
+pub(crate) struct EBSNvmeDevice {
+    pub(crate) device_name: String,
+}
+
+impl From<NvmeIdCtrl> for EBSNvmeDevice {
+    fn from(value: NvmeIdCtrl) -> Self {
+        // The device name is within the first 32 bytes of the vendor specific data, which
+        // is guaranteed to be 1024 bytes long. Trim the resulting string, otherwise udev
+        // doesn't create the correct symlinks.
+        let device_name = String::from_utf8_lossy(&value.vs[0..32]).trim().to_string();
+
+        Self { device_name }
+    }
+}
+
+// Define the nvme_admin_command ioctl function; use `readwrite` because the syscall
+// is defined IOWR.
+// See: https://elixir.bootlin.com/linux/v6.1/source/include/uapi/linux/nvme_ioctl.h#L98
+ioctl_readwrite!(
+    nvme_admin_command,
+    NVME_IOC_MAGIC,
+    NVME_IOC_ADMIN_CMD,
+    NvmePassthruCmd
+);
+
+/// Retrieves the device information for the passed EBS volume
+pub(crate) fn query_ebs_nvme_device(path: &PathBuf) -> Result<EBSNvmeDevice> {
+    let device = fs::File::open(path)
+        .context(error::DeviceOpenSnafu { path })?
+        .into_raw_fd();
+
+    // The structure used to recieve the result of the ioctl syscall, a pointer
+    // to this struct will be send in the ioctl command.
+    let mut ctrl = NvmeIdCtrl::default();
+    let mut cmd = NvmePassthruCmd {
+        opcode: NVME_ADMIN_IDENTIFY,
+        data_len: std::mem::size_of::<NvmeIdCtrl>() as u32,
+        addr: addr_of_mut!(ctrl) as u64,
+        // This is required, otherwise the syscall always returns the same device; this
+        // value corresponds to what Amazon Linux does.
+        // See: https://github.com/amazonlinux/amazon-ec2-utils/blob/09b53f2a20c3e4d227acfb452138fdb672baaa1a/ebsnvme-id#L114
+        cdw10: 1,
+        ..NvmePassthruCmd::default()
+    };
+
+    unsafe {
+        nvme_admin_command(device, &mut cmd).context(error::NvmeIdentifyCtrlSnafu { path })?;
+    }
+
+    let ebs_nvme_device = EBSNvmeDevice::from(ctrl);
+    Ok(ebs_nvme_device)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    /// The size of the NvmeIdCtrl struct
+    const NVME_IDENTIFY_DATA_SIZE: u32 = 4096;
+
+    #[test]
+    // Test that the size of the structure matches the expected size, in case the
+    // structure shape is changed. This is forces developers to update the expected
+    // size of the struct.
+    fn test_nvme_id_ctrl_size() {
```

Comment:
Thank you! I was so relieved when I saw this test. Yes, by all means, let us validate that the huge lump of repr(C) definition actually added up to 4k bytes.